compiler/resolution/src/org/jetbrains/kotlin/effectsystem/adapters/InfoCollector.kt

```diff
@@ -0,0 +1,102 @@
+
+    private fun collectFromClause(clause: ESClause): MutableContextInfo? {
+        val premise = clause.condition
+
+        // Check for non-conditional effects
+        if (premise is ESBooleanConstant && premise.value) {
+            return MutableContextInfo.EMPTY.fire(clause.effect)
+        }
+
+        // Check for information from conditional effects
+        return when (observedEffect.isImplies(clause.effect)) {
+            // observed effect implies clause's effect => clause's effect was fired => clause's condition is true
+            true -> clause.condition.accept(this)
+
+            // Observed effect *may* or *doesn't* implies clause's - no useful information
+            null, false -> null
+        }
+    }
+
+    override fun visitIs(isOperator: ESIs): MutableContextInfo = with(isOperator) {
+        if (isNegated != isInverted) MutableContextInfo.EMPTY.notSubtype(left, type) else MutableContextInfo.EMPTY.subtype(left, type)
+    }
+
+    override fun visitEqual(equal: ESEqual): MutableContextInfo = with(equal) {
+        if (isNegated != isInverted) MutableContextInfo.EMPTY.notEqual(left, right) else MutableContextInfo.EMPTY.equal(left, right)
+    }
+
+    override fun visitAnd(and: ESAnd): MutableContextInfo {
+        val leftInfo = and.left.accept(this)
+        val rightInfo = and.right.accept(this)
+
+        return if (isInverted) leftInfo.or(rightInfo) else leftInfo.and(rightInfo)
+    }
+
+    override fun visitNot(not: ESNot): MutableContextInfo = inverted { not.arg.accept(this) }
+
+    override fun visitOr(or: ESOr): MutableContextInfo {
+        val leftInfo = or.left.accept(this)
+        val rightInfo = or.right.accept(this)
+        return if (isInverted) leftInfo.and(rightInfo) else leftInfo.or(rightInfo)
+    }
+
+    override fun visitVariable(esVariable: ESVariable): MutableContextInfo {
+        throw IllegalStateException("InfoCollector shouldn't visit non-boolean values")
+    }
+
+    override fun visitBooleanVariable(esBooleanVariable: ESBooleanVariable): MutableContextInfo =
+            if (isInverted)
+                MutableContextInfo.EMPTY.equal(esBooleanVariable, false.lift())
+            else
+                MutableContextInfo.EMPTY.equal(esBooleanVariable, true.lift())
+
+    override fun visitBooleanConstant(esBooleanConstant: ESBooleanConstant): MutableContextInfo {
+        throw IllegalStateException("InfoCollector shouldn't visit constants")
+    }
+
+    override fun visitConstant(esConstant: ESConstant): MutableContextInfo {
+        throw IllegalStateException("InfoCollector shouldn't visit constants")
+    }
+
+    override fun visitLambda(esLambda: ESLambda): MutableContextInfo {
+        throw IllegalStateException("InfoCollector shouldn't visit non-boolean values")
+    }
+
+    private fun <R> inverted(block: () -> R): R {
+        isInverted = isInverted.not()
+        val result = block()
+        isInverted = isInverted.not()
```

Comment:
Potentially, this can be fixed by using visitor with data and passing around boolean flag `isInverted` as this data.